### ISSUE TITLE
golang.org/x/net/html Not parsing all Nodes of html string.

### ISSUE THREAD

**Issue by RayfenWindspear** [golang/go#15996]
Please answer these questions before submitting your issue. Thanks!

1. What version of Go are you using (`go version`)?

go version go1.6.1 linux/amd64

2. What operating system and processor architecture are you using (`go env`)?

GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/taeler/go"
GORACE=""
GOROOT="/usr/lib/go-1.6"
GOTOOLDIR="/usr/lib/go-1.6/pkg/tool/linux_amd64"
GO15VENDOREXPERIMENT="1"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="g++"
CGO_ENABLED="1"

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

`package main

import (
	"bytes"
	"fmt"
	"golang.org/x/net/html"
)

var str = `<section id="job-snapshot-section">
    <h2>Job Snapshot</h2>
    <div class="section-body">
                                                
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl10_SnapshotKey">Base Pay</span></strong>
            <span><span id="CBBody_Snapshot__ctl10_SnapshotValue">$150,000.00 - $160,000.00 /Year</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl11_SnapshotKey">Other Pay</span></strong>
            <span><span id="CBBody_Snapshot__ctl11_SnapshotValue">Bonus + Benefits</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl12_SnapshotKey">Employment Type</span></strong>
            <span><span id="CBBody_Snapshot__ctl12_SnapshotValue">Full-Time</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl13_SnapshotKey">Job Type</span></strong>
            <span><span id="CBBody_Snapshot__ctl13_SnapshotValue">Information Technology,  Consultant,  Executive</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl14_SnapshotKey">Education</span></strong>
            <span><span id="CBBody_Snapshot__ctl14_SnapshotValue">Not Specified</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl15_SnapshotKey">Experience</span></strong>
            <span><span id="CBBody_Snapshot__ctl15_SnapshotValue">Not Specified</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl16_SnapshotKey">Manages Others</span></strong>
            <span><span id="CBBody_Snapshot__ctl16_SnapshotValue">Not Specified</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl17_SnapshotKey">Industry</span></strong>
            <span><span id="CBBody_Snapshot__ctl17_SnapshotValue">Employment - Recruiting - Staffing, Computer Software</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl18_SnapshotKey">Required Travel</span></strong>
            <span><span id="CBBody_Snapshot__ctl18_SnapshotValue">Not Specified</span></span>
        </div>                                        
    
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl19_SnapshotKey">Job ID</span></strong>
            <span><span id="CBBody_Snapshot__ctl19_SnapshotValue">46942</span></span>
        </div>                                        
            
    </div>
</section>`

func main() {
	buf := bytes.NewBufferString(str)
	nod, err := html.Parse(buf)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Printf("node: %v\n", nod)
	fmt.Printf("html: %v\n", nod.FirstChild)
	fmt.Printf("head: %v\n", nod.FirstChild.FirstChild)
	fmt.Printf("body: %v\n", nod.FirstChild.LastChild)
	fmt.Printf("section: %v\n", nod.FirstChild.LastChild.FirstChild)
	section := nod.FirstChild.LastChild.FirstChild
	fmt.Printf("h2: %v\n", section.FirstChild)
	fmt.Printf("div: %v\n", section.LastChild)
	fmt.Printf("div: %v\n", section.LastChild.FirstChild)
}
`

4. What did you expect to see?

The last printed div is supposed to have children as per the html string.

5. What did you see instead?

Instead it is parsed as a Text NodeType, and thus has no children, or even data.

**Comment by RayfenWindspear** [comment]
One other thing to note is that calling html.Render will reproduce the missing elements, despite them not being represented as Nodes.

**Comment by kennygrant** [comment]
Nodes also contain TextNodes for their content (including whitespace), which is what you're seeing here. 

Adding a test for section.LastChild.Type to your example shows it is of type TextNode (either with this simpler html or with your example). Removing the whitespace from your example means it works as you expect.

https://play.golang.org/p/uGfE2LTsjg

You can test this with a simpler version of the html, with some whitespace before the last section tag, which becomes the last child node:

`<section><div><div></div></div> </section>`

So I don't think this is a bug?

**Comment by RayfenWindspear** [comment]
Isn't whitespace supposed to be ignored as part of the HTML specification? I'm not sure if this is the right reference, but this portion of the HTML 5 spec is about space separated tokens, which I'm assuming should include tags.

https://html.spec.whatwg.org/multipage/infrastructure.html#space-separated-tokens

**Comment by RayfenWindspear** [comment]
This may be a better resource. It specifically mentions inter-element-whitespace. It states that these Text nodes consisting of only whitespace must be ignored.

https://www.w3.org/TR/html/dom.html#inter-element-whitespace

**Comment by kennygrant** [comment]
Can you reproduce this in a browser using the DOM? If I try this in Chrome or Safari on your snippet or the simpler document (NB whitespace):

`<html><head></head><body><section><div><div></div></div> </section></body></html>`

using a similar query:

`document.firstChild.lastChild.firstChild.lastChild`

I get a textNode with whitespace content, which mirrors what this parser is doing. Obviously on rendering that content the whitespace would be ignored, but it does seem to be in the DOM, and I'm not sure why the spec is implying something browsers don't do there. 

I will leave it to the maintainer to comment further but I hope this helps a little.

**Comment by RayfenWindspear** [comment]
Interesting. I think I see what is going on here. It's not that the browser doesn't support this, it supplies it in a different way. Here is the functionality I have been expecting `document.firstElementChild`, which skips the empty Text Nodes as in the html specification. However, the net/html package only supplies the functionality of the form `document.firstChild`, which does not. 

Given this realization, yes you are right to say that the package isn't malfunctioning. So in effect, this is actually a feature request rather than a bug report.

**Comment by RayfenWindspear** [comment]
Chose not to edit previous post.

So the functionality I am looking for is easily provided by a for loop skipping over Text Nodes. Doing this manually provides me with the data that I am looking for as in this example.

https://play.golang.org/p/ZNKgZATPUI

**Comment by quentinmit** [comment]
@RayfenWindspear You can use this issue for tracking that addition. I don't know how likely it is to be accepted, though, since as you demonstrated it is easy to implement yourself.

/cc @adg @nigeltao Are we likely to add these methods?

**Comment by nigeltao** [comment]
Yes, whitespace is ignored, but at rendering time, not at parse time. kennygrant gave a javascript example, but for a Python example, this program:

```
#!/usr/bin/python

import html5lib

h = '''<section id="job-snapshot-section">
    <h2>Job Snapshot</h2>
    <div class="section-body">
                                                
        <div class="snap-line">
            <strong><span id="CBBody_Snapshot__ctl10_SnapshotKey">Base Pay</span></strong>
            <span><span id="CBBody_Snapshot__ctl10_SnapshotValue">$150,000.00 - $160,000.00 /Year</span></span>
        </div>                                        
    </div>
</section>'''

doc = html5lib.parse(h)
for t in doc.itertext():
    print "[" + t + "]"
```

prints

```
[
    ]
[Job Snapshot]
[
    ]
[
                                                
        ]
[
            ]
[Base Pay]
[
            ]
[$150,000.00 - $160,000.00 /Year]
[
        ]
[                                        
    ]
[
]
```

because those (whitespace) text nodes are part of the DOM tree.

As for a FirstElementChild method, I suppose it's possible, although it's straightforward to provide that as a function (in your package). It doesn't need to be a method (in package html).

Or, if you want to loop only over element children, it's trivial:

```
for c := n.FirstChild; c != nil; c = c.NextSibling {
  if c.Type != html.ElementNode {
    continue
  }
  etc
}
```

so I'm not convinced yet for the need for FirstElementChild and NextElementSibling methods.
